docs/raft.md

```diff
@@ -0,0 +1,137 @@
+# Orchestrator/raft, consensus cluster
+
+![orchestrator HA via raft](images/orchestrator-ha--raft.png)
+
+`orchestratpr/raft` is a deployment setup where several `orchestrator` nodes communicate with each other via `raft` consensus protocol.
+
+`orchestrator/raft` deployments solve both high-availability for `orchestrator` itself as well as solve issues with network isolation, and in particular cross-data-center network partitioning/fencing.
+
+### Very brief overview of traits of raft
+
+By using a consensus protocol the `orchestrator` nodes are able to pick a leader that has _quorum_, implying it is not isolated. For example, consider a `3` node `orchestrator/raft` setup. Normally the three nodes will chat with each other and one will be a stable elected leader. However in face of network partitioning, say node `n1` is partitioned away from nodes `n2` and `n3`, it is guaranteed that the leader will be either `n2` or `n3`. `n1` would not be able to lead because it does not have a quorum (in a `3` node setup the quorum size is `2`; in a `5` node setup the quorum size is `3`)
+
+This turns useful in cross data-center (DC) setups. Assume you set three `orchestrator` nodes, each on its own DC. If one DC gets isolated, it is guaranteed the active `orchestrator` node will be one that has consensus, i.e. operates from outside the isolated DC.
+
+### orchestrator/raft setup technical details
+
+See also: [orchestrator/raft vs. synchronous replication setup](raft-vs-sync-repl.md)
+
+#### Service nodes
+
+You will set up `3` or `5` (recommended raft node count) `orchestrator` nodes. Other numbers are also legitimate but you will want at least `3`.
+
+At this time `orchestrator` nodes to not join dynamically into the cluster. The list of nodes is preconfigured as in:
+
+```json
+  "RaftEnabled": true,
+  "RaftBind": "<IP of this host>",
+  "DefaultRaftPort": 10008,
+  "RaftNodes": [
+    "my.orchestrator.node0",
+    "my.orchestrator.node1",
+    "my.orchestrator.node2",
+  ],
+```
+
+#### Backend DB
+
+Each `orchestrator` node has its own, independent backend database server. This would be either:
+
+- A MySQL backend DB (no replication setup required, but OK if this server has replicas)
+
+  As deployment suggestion, this MySQL server can run on the same `orchestrator` node host.
+
+- A SQLite backend DB. Use:
+```json
+  "BackendDB": "sqlite",
+  "SQLite3DataFile": "/var/lib/orchestrator/orchestrator.db",
+```
+
+`orchestrator` is bundled with `sqlite`, there is no need to install an external dependency.
+
+#### Proxy
+
+You should only send requests to the leader node. To make that happen set up a `HTTP` proxy (e.g HAProxy) on top of the `orchestrator` services.
+
+- Use `/api/leader-check` as health check. At any given time at most one `orchestrator` node will reply with `HTTP 200/OK` to this check; the others will respond with `HTTP 500/Internal server error`.
```

Comment:
Also does this mean that nodes that are not the master also will throw an error if http requests are incorrectly sent to them? Or would it still accept the http requests and cause potential issues? As a future development, would it make sense a node forwards requests to the leader automatically if it's not the leader at the http level?